features/include/pcl/features/impl/good.hpp

```diff
@@ -0,0 +1,454 @@
+    const int idx_i = (int) trunc (b / interval); //inner index
+    if ((idx_o < BinN)
+          and (idx_i < BinN)
+          and (idx_o >= 0)
+          and (idx_i >= 0))
+      ++histogram[idx_o][idx_i];
+  }  
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////
+template<typename PointInT, int BinN> void
+pcl::GOODEstimation<PointInT, BinN>::normalizeHistogram (const std::vector<unsigned int> histogram, std::vector<float> &normalized_histogram)
+{
+  ///compute sumation of all histogram's bins. 
+  float sum_all_bins = std::accumulate (histogram.begin (), histogram.end (), 0);
+  
+  if (sum_all_bins != 0)
+  { 
+    size_t idx = 0; 
+    for(std::vector<unsigned int>::const_iterator it = histogram.begin(); it != histogram.end(); ++it)
+    {
+      normalized_histogram.at (idx) = (*it / sum_all_bins);
+      idx++;
+    }  
+  }
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////
+template<typename PointInT, int BinN> float
+pcl::GOODEstimation<PointInT, BinN>::viewpointEntropy (const std::vector<float> &normalized_histogram)
+{
+  ///NOTE: http://stats.stackexchange.com/questions/66108/why-is-entropy-maximised-when-the-probability-distribution-is-uniform
+  float entropy =0;  
+  for(std::vector<float>::const_iterator it = normalized_histogram.begin(); it != normalized_histogram.end(); ++it)
+  {
+    if (*it != 0)
+    {
+      float entropy_tmp = *it * log2 (*it);
+      entropy += entropy_tmp;
+    }
+  }  
+  return (-entropy); 
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////
+template<typename PointInT, int BinN> int 
+pcl::GOODEstimation<PointInT, BinN>::findMaxViewPointEntropy (const std::vector<float> &view_point_entropy)
+{
+  int index = 0;
+  std::vector<float>::const_iterator it;
+  it = std::max_element (view_point_entropy.begin(), view_point_entropy.end());
+  index = it - view_point_entropy.begin ();
+  return (index);
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////
+template<typename PointInT, int BinN> float
+pcl::GOODEstimation<PointInT, BinN>::varianceOfHistogram (const std::vector<float> &histogram)
+{
+  ///NOTE: https://people.richland.edu/james/lecture/m170/ch06-prb.html
+  ///NOTE: http://www.stat.yale.edu/Courses/1997-98/101/rvmnvar.htm  
+  
+  float mean = 0;
+  float variance = 0;
+  for (size_t i = 0; i < histogram.size(); i++)
+  {
+    mean += (i+1) * histogram.at(i);
+  }
+  
+  for (size_t i = 0; i < histogram.size(); i++)
+  {
+    variance += std::pow ((i+1)-mean , 2) * histogram.at(i);
+  }  
+  return (variance);  
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////
+template<typename PointInT, int BinN> void
+pcl::GOODEstimation<PointInT, BinN>::objectViewHistogram (int maximum_entropy_index, 
```

Comment:
`maximum_entropy_index` can only be 0 to 2, which means that this is actually a `Projection` type.